### PR TITLE
fix(gatsby): queue fs writes so event loop not overwhelmed on large sites

### PR DESCRIPTION
**Pull request by KyleAMathews**
Dropped write time for 100k create-pages benchmark from 69s -> 15s for a ~78% decrease

### CODE REVIEW

File: packages/gatsby/src/utils/page-data.ts [hunk 1]
```diff
@@ -162,7 +163,7 @@ export async function flush(): Promise<void> {
 
         if (hasFlag(query.dirty, FLAG_DIRTY_NEW_PAGE)) {
           // query results are not written yet
-          continue
+          cb(null, true)
```
Comment (KyleAMathews) [inline packages/gatsby/src/utils/page-data.ts:166]: @pieh any idea why this isn't working?

Comment (vladar) [reply packages/gatsby/src/utils/page-data.ts:166]: Should you also add `return` after this line (as we want to skip the remaining part of this function)?

Comment (pieh) [reply packages/gatsby/src/utils/page-data.ts:166]: What Vlad said, or keep part of the loop as-is now, and only make async part of the loop use queue - so change https://github.com/gatsbyjs/gatsby/blob/bca9518abd0f3f4560c25b24fda04955baaf7528/packages/gatsby/src/utils/page-data.ts#L172-L185

to push to queue and queue handler only does that part

Comment (KyleAMathews) [reply packages/gatsby/src/utils/page-data.ts:166]: 🤦 yes, returning helps

File: packages/gatsby/src/utils/page-data.ts [hunk 2]
```diff
@@ -134,7 +135,7 @@ export async function flush(): Promise<void> {
 
   const pagesToWrite = pagePaths.values()
 
-  for (const pagePath of pagesToWrite) {
+  const flushQueue = fastq(async (pagePath, cb) => {
```
Comment (vladar) [inline packages/gatsby/src/utils/page-data.ts:138]:
```suggestion
  const flushQueue = fastq((pagePath, cb) => {
```
I don't think we need this (`fastq` doesn't seem to use promises)

Comment (KyleAMathews) [reply packages/gatsby/src/utils/page-data.ts:138]: we do for the await

Comment (vladar) [reply packages/gatsby/src/utils/page-data.ts:138]: oh, right

**Review by pieh — APPROVED** [review]
😍  thanks!

**Pull request merged**